cmd/cli/main.go

```diff
@@ -14,6 +14,155 @@
 
 package main
 
-func main() {
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/spf13/cobra"
+
+	"antrea.io/resource-auditing/pkg/types"
+)
+
+// tag flags
+var tagAuthor, tagEmail string
+
+// rollback flags
+var rollbackTag, rollbackSHA string
+
+var commandName = path.Base(os.Args[0])
+
+const port = "8080"
+
+var rootCmd = &cobra.Command{
+	Use:  commandName,
+	Long: commandName + " is the command line tool for managing the auditing resource repository",
+}
+
+var tagCmd = &cobra.Command{
+	Use:   "tag create tag_name commit_sha [-a author] [-e email]\n   or: tag delete tag_name",
+	Short: "tags commits in the repository",
+	Args: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 2 {
+			return fmt.Errorf("too few args")
+		}
+		if args[0] == "create" {
+			if len(args) != 3 {
+				return fmt.Errorf("unexpected number of args for tag create")
+			}
+		} else if args[0] == "delete" {
+			if len(args) != 2 {
+				return fmt.Errorf("unexpected number of args for tag delete")
+			}
+		} else {
+			return fmt.Errorf("unsupported keyword (not create or delete)")
+		}
+		return nil
+	},
+	Run: runTag,
+}
+
+var rollbackCmd = &cobra.Command{
+	Use:   "rollback -t tag_name | -s commit_sha",
+	Short: "rollback to the specified commit by tag name or SHA",
+	Args: func(cmd *cobra.Command, args []string) error {
+		if len(args) != 0 {
+			return fmt.Errorf("unexpected number of args for rollback")
+		}
+		if (rollbackTag != "") == (rollbackSHA != "") {
+			return fmt.Errorf("must specify exactly one of -t or -s")
+		}
+		return nil
+	},
+	Run: runRollback,
+}
+
+func runTag(cmd *cobra.Command, args []string) {
+	url := "http://localhost:" + port + "/tag"
```

Comment:
please make this a CLI flag for both commands (`--server-addr`) which can default to that value

I seem to remember that there is a way to make it a shared flag for multiple commands, using a group or something. Could you check?

Reply:
Didn't find anything related to groups but found this which says we can use the viper library to bind the same flag name to two different subcommands dynamically at runtime: https://github.com/spf13/viper/issues/233

But I simply just made the two commands use the same global flag, which works too - let me know if that's sufficient.